solidity/src/Base7683.sol

```diff
@@ -0,0 +1,410 @@
+    )
+        internal
+    {
+        OrderData memory orderData = OrderEncoder.decode(order.orderData);
+
+        PERMIT2.permitWitnessTransferFrom(
+            ISignatureTransfer.PermitTransferFrom({
+                permitted: ISignatureTransfer.TokenPermissions({
+                    token: TypeCasts.bytes32ToAddress(orderData.inputToken),
+                    amount: orderData.amountIn
+                }),
+                nonce: order.nonce,
+                deadline: order.openDeadline
+            }),
+            ISignatureTransfer.SignatureTransferDetails({ to: receiver, requestedAmount: orderData.amountIn }),
+            order.user,
+            witnessHash(order),
+            witnessTypeString,
+            signature
+        );
+    }
+
+    function _settleOrder(bytes32 _orderId, bytes32 _receiver, uint32 _settlingDomain) internal {
+        OrderData memory orderData = orders[_orderId];
+
+        if (orderData.destinationDomain != _settlingDomain) revert InvalidDomain();
+        if (orderStatus[_orderId] != OrderStatus.OPENED) revert InvalidOrderStatus();
+
+        orderStatus[_orderId] = OrderStatus.SETTLED;
+
+        address receiver = TypeCasts.bytes32ToAddress(_receiver);
+
+        emit Settled(_orderId, receiver);
+
+        IERC20(TypeCasts.bytes32ToAddress(orderData.inputToken)).safeTransfer(
+            receiver, orderData.amountIn
+        );
+    }
+
+    function _refundOrder(bytes32 _orderId, uint32 _refundingDomain) internal {
+        OrderData memory orderData = orders[_orderId];
+
+        if (orderData.destinationDomain != _refundingDomain) revert InvalidDomain();
+        if (orderStatus[_orderId] != OrderStatus.OPENED) revert InvalidOrderStatus();
+
+        orderStatus[_orderId] = OrderStatus.REFUNDED;
+
+        address orderSender = TypeCasts.bytes32ToAddress(orderData.sender);
+
+        emit Refunded(_orderId, orderSender);
+
+        IERC20(TypeCasts.bytes32ToAddress(orderData.inputToken)).safeTransfer(
+            orderSender, orderData.amountIn
+        );
+    }
+
+    /**
+     * @dev This function is called during `settle` to handle the settlement of the orders, it is meant to be
+     * implemented by the inheriting contract with specific settlement logic. i.e. sending a cross-chain message
+    */
+    function _handleSettlement(bytes32[] memory _orderIds, bytes32[] memory _receivers) internal virtual;
+
+    /**
+     * @dev This function is called during `settle` to handle the settlement of the orders, it is meant to be
+     * implemented by the inheriting contract with specific settlement logic. i.e. sending a cross-chain message
+    */
+    function _handleRefund(bytes32[] memory _orderIds) internal virtual;
+
+    /**
+     * @dev To be implemented by the inheriting contract with specific logic, the address of its remote counterpart and
+     * revert if it does not exist
+    */
+    function _mustHaveRemoteCounterpart(uint32 _domain) internal view virtual returns (bytes32);
```

Comment:
This is an interesting question, my gut reaction would have been that contracts don't have to do this invariant checking, its up to the solver to know whether this contract supports settlement of a particular origin (same as the refund). Some systems like across dont have this notion of a remoteCounterpart, everything settles on L1, even if the order origin is from elsewhere.

I do like the idea of overridable hooks for custom logic for invariant checking in fill/open/refund, but probably we'll get a better idea for hooks once we have more implementations of settlement